File: frontend/src/services/currencyService.ts
import {
  ConvertCurrency,
  GetSupportedCurrencies,
  GetExchangeRatesForCurrency,
  ClearCurrencyCache,
} from "../../wailsjs/go/main/App";
import { CurrencyConversionCache } from "../utils/currencyCache";

export interface ExchangeRates {
  [currency: string]: number;
}

export interface CurrencyConfig {
  code: string;
  symbol: string;
  name: string;
  rate: number; // currency conversion rate relative to BRL
}

export class CurrencyService {
  private static instance: CurrencyService;
  private supportedCurrencies: CurrencyConfig[] = [];
  private exchangeRates: ExchangeRates = {};
  private lastFetchTime: number = 0;
  private readonly CACHE_DURATION = 30 * 60 * 1000;
  private conversionCache: CurrencyConversionCache;

  private constructor() {
    this.initializeBasicData();
    this.loadSupportedCurrencies();
    this.conversionCache = CurrencyConversionCache.getInstance();
  }

  private initializeBasicData(): void {
    this.supportedCurrencies = [
      { code: "BRL", symbol: "R$", name: "Real Brasileiro", rate: 1.0 },
      { code: "USD", symbol: "$", name: "US Dollar", rate: 0.2 },
      { code: "EUR", symbol: "€", name: "Euro", rate: 0.18 },
      { code: "GBP", symbol: "£", name: "British Pound", rate: 0.15 },
      { code: "JPY", symbol: "¥", name: "Japanese Yen", rate: 30.0 },
      { code: "CAD", symbol: "C$", name: "Canadian Dollar", rate: 0.27 },
      { code: "AUD", symbol: "A$", name: "Australian Dollar", rate: 0.3 },
      { code: "CHF", symbol: "CHF", name: "Swiss Franc", rate: 0.18 },
      { code: "CNY", symbol: "¥", name: "Chinese Yuan", rate: 1.45 },
      { code: "INR", symbol: "₹", name: "Indian Rupee", rate: 16.8 },
    ];

    this.exchangeRates = {
      BRL: 1.0,
      USD: 0.2,
      EUR: 0.18,
      GBP: 0.15,
      JPY: 30.0,
      CAD: 0.27,
      AUD: 0.3,
      CHF: 0.18,
      CNY: 1.45,
      INR: 16.8,
    };

    this.lastFetchTime = Date.now();
  }

  public static getInstance(): CurrencyService {
    if (!CurrencyService.instance) {
      CurrencyService.instance = new CurrencyService();
    }
    return CurrencyService.instance;
  }

  private async loadSupportedCurrencies(): Promise<void> {
    try {
      const response = await GetSupportedCurrencies();
      this.supportedCurrencies = response.currencies.map(currency => ({
        code: currency.code,
        symbol: currency.symbol,
        name: currency.name,
        rate: this.exchangeRates[currency.code] || 1.0,
      }));

      await this.updateExchangeRatesFromBackend();
    } catch (error) {
      console.error("Error loading supported currencies:", error);
    }
  }

  private async updateExchangeRatesFromBackend(): Promise<void> {
    try {
      const response = await GetExchangeRatesForCurrency("BRL");
      if (response.rates) {
        this.exchangeRates = { BRL: 1.0, ...response.rates };

        this.supportedCurrencies = this.supportedCurrencies.map(currency => ({
          ...currency,
          rate: this.exchangeRates[currency.code] || currency.rate,
        }));

        this.lastFetchTime = Date.now();
      }
    } catch (error) {
      console.error("Error updating exchange rates:", error);
    }
  }

  private async ensureRatesUpdated(): Promise<void> {
    const now = Date.now();
    if (now - this.lastFetchTime > this.CACHE_DURATION) {
      await this.updateExchangeRatesFromBackend();
    }
  }

  public async convertFromBRL(
    amount: number,
    toCurrency: string
  ): Promise<number> {
    if (toCurrency === "BRL") return amount;

    const cacheKey = { amount, fromCurrency: "BRL", toCurrency };
    const cachedResult = this.conversionCache.get(cacheKey);

    if (cachedResult !== null) {
      this.conversionCache.recordHit();
      return cachedResult;
    }

    this.conversionCache.recordMiss();

    try {
      await this.ensureRatesUpdated();

      const response = await ConvertCurrency({
        amount,
        fromCurrency: "BRL",
        toCurrency,
      });

      this.conversionCache.set(
        cacheKey,
        response.convertedAmount,
        response.exchangeRate
      );

      return response.convertedAmount;
    } catch (error) {
      console.error(`Error converting from BRL to ${toCurrency}:`, error);

      const rate = this.exchangeRates[toCurrency];
      if (rate) {
        const result = amount * rate;
        this.conversionCache.set(cacheKey, result, rate);
        return result;
      }

      return amount;
    }
  }

  public async convertToBRL(
    amount: number,
    fromCurrency: string
  ): Promise<number> {
    if (fromCurrency === "BRL") return amount;

    const cacheKey = { amount, fromCurrency, toCurrency: "BRL" };
    const cachedResult = this.conversionCache.get(cacheKey);

    if (cachedResult !== null) {
      this.conversionCache.recordHit();
      return cachedResult;
    }

    this.conversionCache.recordMiss();

    try {
      await this.ensureRatesUpdated();

      const response = await ConvertCurrency({
        amount,
        fromCurrency,
        toCurrency: "BRL",
      });

      this.conversionCache.set(
        cacheKey,
        response.convertedAmount,
        response.exchangeRate
      );

      return response.convertedAmount;
    } catch (error) {
      console.error(`Error converting from ${fromCurrency} to BRL:`, error);

      const rate = this.exchangeRates[fromCurrency];
      if (rate) {
        const result = amount / rate;
        this.conversionCache.set(cacheKey, result, 1 / rate);
        return result;
      }

      return amount;
    }
  }

  public async convertCurrency(
    amount: number,
    fromCurrency: string,
    toCurrency: string
  ): Promise<number> {
    if (fromCurrency === toCurrency) return amount;

    const cacheKey = { amount, fromCurrency, toCurrency };
    const cachedResult = this.conversionCache.get(cacheKey);

    if (cachedResult !== null) {
      this.conversionCache.recordHit();
      return cachedResult;
    }

    this.conversionCache.recordMiss();

    try {
      const response = await ConvertCurrency({
        amount,
        fromCurrency,
        toCurrency,
      });

      this.conversionCache.set(
        cacheKey,
        response.convertedAmount,
        response.exchangeRate
      );

      return response.convertedAmount;
    } catch (error) {
      console.error(
        `Error converting from ${fromCurrency} to ${toCurrency}:`,
        error
      );
      return amount;
    }
  }

  public formatCurrency(
    amount: number,
    currency: string,
    locale?: string
  ): string {
    const config = this.getCurrencyConfig(currency);
    if (!config) {
      return `${amount.toFixed(2)}`;
    }

    const localeMap: Record<string, string> = {
      BRL: "pt-BR",
      USD: "en-US",
      EUR: "de-DE",
      GBP: "en-GB",
      JPY: "ja-JP",
      CAD: "en-CA",
      AUD: "en-AU",
      CHF: "de-CH",
      CNY: "zh-CN",
      INR: "en-IN",
    };

    const targetLocale = locale || localeMap[currency] || "en-US";

    try {
      return new Intl.NumberFormat(targetLocale, {
        style: "currency",
        currency: currency,
      }).format(amount);
    } catch {
      return `${config.symbol} ${amount.toFixed(2)}`;
    }
  }

  public async refreshExchangeRates(): Promise<void> {
    try {
      await ClearCurrencyCache();
      this.conversionCache.clear();
      await this.updateExchangeRatesFromBackend();
    } catch (error) {
      console.error("Error refreshing exchange rates:", error);
    }
  }

  public getCurrencyConfig(currency: string): CurrencyConfig | null {
    return this.supportedCurrencies.find(c => c.code === currency) || null;
  }

  public getSupportedCurrencies(): CurrencyConfig[] {
    return this.supportedCurrencies;
  }

  public getExchangeRates(): ExchangeRates {
    return { ...this.exchangeRates };
  }

  public hasValidData(): boolean {
    return Object.keys(this.exchangeRates).length > 0;
  }

  public getLastFetchTime(): Date | null {
    return this.lastFetchTime > 0 ? new Date(this.lastFetchTime) : null;
  }

  public getCacheStats() {
    return this.conversionCache.getStats();
  }

  public async preloadCommonConversions(): Promise<void> {
    const currencies = this.supportedCurrencies.map(c => c.code);
    const commonAmounts = [1, 10, 50, 100, 500, 1000];

    await this.conversionCache.preloadCommonConversions(
      commonAmounts,
      currencies,
      this.convertCurrency.bind(this)
    );
  }
}

File: core/services/currency_service.go
package service

import (
	"context"
	"encoding/json"
	"fmt"
	"io"
	"log"
	"net/http"
	"strings"
	"sync"
	"time"

	"product-management-app/core/dto"
)

type CurrencyService struct {
	ctx                 context.Context
	httpClient          *http.Client
	cachedRates         map[string]map[string]float64 // [baseCurrency][targetCurrency]rate
	cacheExpiry         map[string]time.Time          // [baseCurrency]expiryTime
	cacheMutex          sync.RWMutex
	cacheTimeout        time.Duration
	supportedCurrencies map[string]dto.CurrencyInfo
}

const (
	primaryAPIURL  = "https://cdn.jsdelivr.net/npm/@fawazahmed0/currency-api@latest/v1/currencies"
	fallbackAPIURL = "https://latest.currency-api.pages.dev/v1/currencies"

	defaultCacheTimeout = 30 * time.Minute

	httpTimeout = 10 * time.Second
)

func NewCurrencyService(ctx context.Context) *CurrencyService {
	return &CurrencyService{
		ctx:                 ctx,
		httpClient:          &http.Client{Timeout: httpTimeout},
		cachedRates:         make(map[string]map[string]float64),
		cacheExpiry:         make(map[string]time.Time),
		cacheTimeout:        defaultCacheTimeout,
		supportedCurrencies: initSupportedCurrencies(),
	}
}

func (cs *CurrencyService) logInfo(message string) {
	// Para standalone testing, apenas usa log padrão
	log.Printf("INFO: %s", message)
}

func (cs *CurrencyService) logWarning(message string) {
	log.Printf("WARNING: %s", message)
}

func (cs *CurrencyService) logError(message string) {
	log.Printf("ERROR: %s", message)
}

func initSupportedCurrencies() map[string]dto.CurrencyInfo {
	return map[string]dto.CurrencyInfo{
		"BRL": {Code: "BRL", Symbol: "R$", Name: "Brazilian Real"},
		"USD": {Code: "USD", Symbol: "$", Name: "US Dollar"},
		"EUR": {Code: "EUR", Symbol: "€", Name: "Euro"},
		"GBP": {Code: "GBP", Symbol: "£", Name: "British Pound"},
		"JPY": {Code: "JPY", Symbol: "¥", Name: "Japanese Yen"},
		"CAD": {Code: "CAD", Symbol: "C$", Name: "Canadian Dollar"},
		"AUD": {Code: "AUD", Symbol: "A$", Name: "Australian Dollar"},
		"CHF": {Code: "CHF", Symbol: "CHF", Name: "Swiss Franc"},
		"CNY": {Code: "CNY", Symbol: "¥", Name: "Chinese Yuan"},
		"INR": {Code: "INR", Symbol: "₹", Name: "Indian Rupee"},
	}
}

func (cs *CurrencyService) fetchExchangeRates(baseCurrency string) (map[string]float64, error) {
	baseCurrency = strings.ToLower(baseCurrency)

	rates, err := cs.fetchFromURL(fmt.Sprintf("%s/%s.json", primaryAPIURL, baseCurrency))
	if err != nil {
		cs.logWarning(fmt.Sprintf("Primary API failed for %s: %v, trying fallback", baseCurrency, err))

		rates, err = cs.fetchFromURL(fmt.Sprintf("%s/%s.json", fallbackAPIURL, baseCurrency))
		if err != nil {
			cs.logError(fmt.Sprintf("Both APIs failed for %s: %v", baseCurrency, err))
			return nil, fmt.Errorf("failed to fetch exchange rates from both APIs: %v", err)
		}
	}

	cs.logInfo(fmt.Sprintf("Successfully fetched exchange rates for %s", baseCurrency))
	return rates, nil
}

func (cs *CurrencyService) fetchFromURL(url string) (map[string]float64, error) {
	cs.logInfo(fmt.Sprintf("Fetching exchange rates from: %s", url))

	resp, err := cs.httpClient.Get(url)
	if err != nil {
		return nil, fmt.Errorf("HTTP request failed: %v", err)
	}
	defer resp.Body.Close()

	if resp.StatusCode != http.StatusOK {
		return nil, fmt.Errorf("API returned status code: %d", resp.StatusCode)
	}

	body, err := io.ReadAll(resp.Body)
	if err != nil {
		return nil, fmt.Errorf("failed to read response body: %v", err)
	}

	var apiResponse map[string]interface{}
	if err := json.Unmarshal(body, &apiResponse); err != nil {
		return nil, fmt.Errorf("failed to parse JSON response: %v", err)
	}

	rates := make(map[string]float64)

	for key, value := range apiResponse {
		if key != "date" {
			if ratesMap, ok := value.(map[string]interface{}); ok {
				for currency, rate := range ratesMap {
					if rateFloat, ok := rate.(float64); ok {
						rates[strings.ToUpper(currency)] = rateFloat
					}
				}
			}
		}
	}

	if len(rates) == 0 {
		return nil, fmt.Errorf("no exchange rates found in API response")
	}

	return rates, nil
}

func (cs *CurrencyService) getRatesFromCache(baseCurrency string) (map[string]float64, bool) {
	cs.cacheMutex.RLock()
	defer cs.cacheMutex.RUnlock()

	if expiry, exists := cs.cacheExpiry[baseCurrency]; exists && time.Now().Before(expiry) {
		if rates, exists := cs.cachedRates[baseCurrency]; exists {
			cs.logInfo(fmt.Sprintf("Using cached rates for %s", baseCurrency))
			return rates, true
		}
	}

	return nil, false
}

func (cs *CurrencyService) saveRatesToCache(baseCurrency string, rates map[string]float64) {
	cs.cacheMutex.Lock()
	defer cs.cacheMutex.Unlock()

	cs.cachedRates[baseCurrency] = rates
	cs.cacheExpiry[baseCurrency] = time.Now().Add(cs.cacheTimeout)

	cs.logInfo(fmt.Sprintf("Cached rates for %s until %v", baseCurrency, cs.cacheExpiry[baseCurrency]))
}

func (cs *CurrencyService) getExchangeRate(fromCurrency, toCurrency string) (float64, error) {
	fromCurrency = strings.ToUpper(fromCurrency)
	toCurrency = strings.ToUpper(toCurrency)

	if fromCurrency == toCurrency {
		return 1.0, nil
	}

	if rates, found := cs.getRatesFromCache(fromCurrency); found {
		if rate, exists := rates[toCurrency]; exists {
			return rate, nil
		}
	}

	rates, err := cs.fetchExchangeRates(fromCurrency)
	if err != nil {
		return 0, err
	}

	cs.saveRatesToCache(fromCurrency, rates)

	if rate, exists := rates[toCurrency]; exists {
		return rate, nil
	}

	return 0, fmt.Errorf("exchange rate not found for %s to %s", fromCurrency, toCurrency)
}

func (cs *CurrencyService) ConvertCurrency(request dto.CurrencyConversionRequest) (*dto.CurrencyConversionResponse, error) {
	cs.logInfo(fmt.Sprintf("Converting %.2f %s to %s", request.Amount, request.FromCurrency, request.ToCurrency))

	if request.Amount < 0 {
		return nil, fmt.Errorf("amount must be positive")
	}

	rate, err := cs.getExchangeRate(request.FromCurrency, request.ToCurrency)
	if err != nil {
		return nil, err
	}

	convertedAmount := request.Amount * rate

	response := &dto.CurrencyConversionResponse{
		Amount:          request.Amount,
		FromCurrency:    strings.ToUpper(request.FromCurrency),
		ToCurrency:      strings.ToUpper(request.ToCurrency),
		ConvertedAmount: convertedAmount,
		ExchangeRate:    rate,
		ConversionDate:  time.Now(),
	}

	cs.logInfo(fmt.Sprintf("Conversion successful: %.2f %s = %.2f %s (rate: %.6f)",
		request.Amount, response.FromCurrency, convertedAmount, response.ToCurrency, rate))

	return response, nil
}

func (cs *CurrencyService) GetSupportedCurrencies() *dto.SupportedCurrenciesResponse {
	currencies := make([]dto.CurrencyInfo, 0, len(cs.supportedCurrencies))
	for _, currency := range cs.supportedCurrencies {
		currencies = append(currencies, currency)
	}

	return &dto.SupportedCurrenciesResponse{
		Currencies: currencies,
	}
}

func (cs *CurrencyService) GetExchangeRatesForCurrency(baseCurrency string) (*dto.CurrencyRatesResponse, error) {
	baseCurrency = strings.ToUpper(baseCurrency)
	cs.logInfo(fmt.Sprintf("Getting all exchange rates for %s", baseCurrency))

	if rates, found := cs.getRatesFromCache(baseCurrency); found {
		return &dto.CurrencyRatesResponse{
			Date:  time.Now().Format("2006-01-02"),
			Base:  baseCurrency,
			Rates: rates,
		}, nil
	}

	rates, err := cs.fetchExchangeRates(baseCurrency)
	if err != nil {
		return nil, err
	}

	cs.saveRatesToCache(baseCurrency, rates)

	return &dto.CurrencyRatesResponse{
		Date:  time.Now().Format("2006-01-02"),
		Base:  baseCurrency,
		Rates: rates,
	}, nil
}

func (cs *CurrencyService) ClearCache() {
	cs.cacheMutex.Lock()
	defer cs.cacheMutex.Unlock()

	cs.cachedRates = make(map[string]map[string]float64)
	cs.cacheExpiry = make(map[string]time.Time)

	cs.logInfo("Currency exchange rates cache cleared")
}
